OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -303,23 +275,59 @@ public bool CanEnterCell(World world, Actor self, CPos cell, out int movementCos
 		}
 
 		public int GetInitialFacing() { return InitialFacing; }
+
+		/// <summary>
+		/// Checks whether the actor "self" collides against the actor "actor" or whether
+		/// it has to ignore the "ignoreActor"
+		/// </summary>
+		bool Collides(Actor self, Actor actor, Actor ignoreActor, bool canIgnoreMovingAllies)
+		{
+			if (actor == ignoreActor)
+				return false;
+
+			// Neutral/enemy units are blockers. Allied units that are moving are not blockers.
+			if (canIgnoreMovingAllies && self.IsAlliedWith(actor) && IsMovingInMyDirection(self, actor))
+				return false;
+
+			// Non-sharable unit can enter a cell with shareable units only if it can crush all of them.
+			if (CanBeCrushedBy(self, actor))
+				return false;
+
+			return true;
+		}
+
+		bool CanBeCrushedBy(Actor crusher, Actor actor)
+		{
+			if (Crushes == null || Crushes.Length == 0)
+				return false;
+
+			var crushables = actor.TraitsImplementing<ICrushable>();
+			if (!crushables.Any() || crushables.Any(crushable => !crushable.CrushableBy(Crushes, crusher.Owner)))
+				return false;
+
+			return true;
+		}
 	}
 
-	public class Mobile : IIssueOrder, IResolveOrder, IOrderVoice, IPositionable, IMove, IFacing, ISync, INotifyAddedToWorld, INotifyRemovedFromWorld, INotifyBlockingMove
+	public class Mobile : IMove, IPositionable, IFacing, IIssueOrder, IResolveOrder, IOrderVoice, ISync, INotifyAddedToWorld, INotifyRemovedFromWorld, INotifyBlockingMove
 	{
 		const int AverageTicksBeforePathing = 5;
 		const int SpreadTicksBeforePathing = 5;
+
 		internal int TicksBeforePathing = 0;
 
 		readonly Actor self;
+
 		public readonly MobileInfo Info;
+
 		public bool IsMoving { get; set; }
 
 		int facing;
 		CPos fromCell, toCell;
 		public SubCell FromSubCell, ToSubCell;
 
-		[Sync] public int Facing
+		[Sync]
+		public int Facing
```

Comment:
Don't commit this formatting change (it'll be inconsistent with the rest of the codebase).